Add the ability to track any qualities, not just favours

AnteIndustrial opened the pull request:
Add the ability to track any qualities, not just favours

AnteIndustrial committed:
FEATURE: Expand on favour trackers to allow any quality or item to be tracked

lensvol commented on src/fixers/index.ts:
```diff
@@ -43,6 +44,7 @@ export default [
     ThingSortFixer,
     AsceticModeFixer,
     FavourTrackerFixer,
+    MiscTracker,
```
The convention is to name classes ending with "Fixer".

lensvol commented on src/fixers/misc_tracker.ts:
```diff
@@ -0,0 +1,348 @@
+        });
+    }
+
+    private stringSorter(): ((a: string, b: string) => number) | undefined {
```
I believe that could be replaced with `const stringSorter = (s1, s2) => (s1 > s2 ? -1 : 1)`

lensvol commented on src/fixers/misc_tracker.ts:
```diff
@@ -0,0 +1,348 @@
+                const newQuality = { name: trackName, category: trackCategory, currentValue: trackCurrent, targetValue: trackNumber, image: trackImage };
+                console.log("before add")
+                console.log(this.miscQualities)
+                console.log(this.currentSettings)
```
If you output debug information, then please use the appropriate `debug()` function supplied by the extension itself. It will both set the appropriate reporting level and append the prefix annotating that the message belongs to this extension.

AnteIndustrial replied:
Woops, these should have all been removed once I got it working

lensvol reviewed (changes requested):
Please replace your current method of arranging qualities to be tracked with an appropriate setting, leaving its handling to the extension itself. This will allow a lot of code here to be removed and your custom tracker element to be replaced with the same used in a `FavourTrackerFixer`, allowing code to be de-duplicated.

AnteIndustrial committed:
Moving tracking config to the settings page, removing logging that wasn't meant to be there, making some of the suggested changes

AnteIndustrial committed:
A bunch of bugfixes, up and down arrows, making buttons smaller so it looks a little nicer

AnteIndustrial committed:
fixing a bug that reset trackers every time I reloaded the page

(cherry picked from commit 5dc1b3ad60de5ab6b7667e13f90d29a32b26b28d)

lensvol commented:
While I am impressed with the amount of progress made here, I am still not sure that I am happy with how it looks right now:

* There is too much happening in the settings and even game state leaks there now. It was never intended for this and current implementation also adds unnecessary coupling between unrelated entities. Moving it all to a pop-up modal that gets called from the tracker panel itself may make more sense. In this way, the toggle for the feature itself would still be where it is expected (settings) while arbitrarily complex logic for handling quality management will happen in a dedicated space.
* At the current moment `onCharacterDataLoaded` is unreliable due to the issue of the extension sometimes loading too late and missing the `/myself` response that contains data about all the qualities of the current character. Not to mention that it will not work for the transient tracking qualities like e.g. `Fivefold Devotion` which only exists in specific situations, so you'll have to be at a specific place to be able to add it to the list. What I can suggest is to download data about all the qualities (name, image, whether it is discrete), process it into JS code, and store it alongside other static data in an extension.
* Having a target value for the quality may make sense for some of them, but useless for others like "Airs of..." since they change non-linearly. Could it be done in a way that makes specific targets optional?
* Merging the existing Favour tracker and yours makes sense, but makes it non-trivial to understand where those Favours are coming from. Maybe they should also be prepopulated in the list of tracked qualities?

Other than that - great work!

AnteIndustrial commented:
Yes, gamestate coming into the settings, and settings doing a bunch of html stuff, definitely is not ideal. I can give a modal a shot.

I don't really like the idea of downloading all qualities to prepopulate, because then it would need a new download every time a new item is added to the game, or a new area, or a new event. Also the wiki doesn't track paid content, so stuff like the summer nursery and every new exceptional story would still need to pull data from the game at runtime. I also don't think it's necessary: you can type in whatever quality you want, and if the game hasn't seen it yet it will get a question mark icon, but as soon as you get one it will pull the new icon and track it normally. All a user needs to do is type it in correctly, which is not ideal but seems better than constantly updating a list that's never going to be complete anyway.

I'll have a look at making the target thing better. Yeah, I've noticed it's not great for Airs, especially if you want, say, Airs 23-45, you can track 45 but how do you set the minimum? And in Port Carnelian you want to target just over 105 or just over 176, which this isn't great for either. I guess if I'm moving the editing thing to a modal, I'll have more room that when it's on the side panel or in settings, and more room means I could add more settings like 'target', 'target range', 'multiple targets'.

Hmm, I could make it so that if settings.trackedQualities === null, I put in favours, but if it's === "" I leave it empty? That way when someone first changes over from favours to qualities it will prepopulate it with favours, but if they delete them all the favours don't come back?

AnteIndustrial committed:
took tracking out of settings, game state is no longer linked to settings. Changing tracked qualities now happens in a popup modal.
If settings.trackedQualities is null, it gets populated with favours. This should happen the first time you use it. If you then delete all tracked qualities, settings.trackedQualities is an empty list, not null, so it won't get repopulated with defaults.

AnteIndustrial commented:
> While I am impressed with the amount of progress made here, I am still not sure that I am happy with how it looks right now:

- I've moved config out of settings and into a modal. It works much nicer now, good call on that.
- I haven't done anything about this yet
- Or this
- Favours don't come back if you delete them any more. The first time you turn the feature on it will add all favours, but if you delete them from tracking they won't come back.
I also made a few changes to keep things updated better - if you track something that you currently have 0 of it will be updated as soon as you get 1, images and progress bars are updated more reliably, changes on the edit modal are reflected immediately on the tracking panel.

There's now an awful lot of 'build some html' code, is there a preferred way to handle this? Like, I could move that code to a TrackedQualitiesHtmlBuilder or something? I'm not sure whether that would be a good idea.

Entirely unrelated, do you think the option 'Protect certain 'dangerous' branches with a two-step confirmation process' could become user-configurable? For instance, I was hunting parabolan animals a while ago, and would have liked to block the 'abandon the hunt' options to protect me from misclicks. Maybe a button on the storylet to block that option, or the settings page accepts a list of IDs, or something?

AnteIndustrial committed:
missed one bug

AnteIndustrial commented:
> * At the current moment `onCharacterDataLoaded` is unreliable due to the issue of the extension sometimes loading too late and missing the `/myself` response that contains data about all the qualities of the current character.
I'm not sure what to do about this. I looked at adding an API call to /myself, but then I read the devs saying extensions calling their servers isn't allowed. I think in this case the best option is a message saying to refresh the page, do you agree?

> * Having a target value for the quality may make sense for some of them, but useless for others like "Airs of..." since they change non-linearly. 
I was looking at Airs of London because of this comment, but it seems to work differently, more like a hidden quality, it doesn't get returned from /myself.
Which leads into another problem, which is that hidden qualities can't be easily tracked. If I check /myself for The Airs of London, or whatever other hidden quality, there's no entry for it, but I can't just disregard that because no entry is exactly how the game tells me the value == 0. I could theoretically get a list of all hidden qualities, and not set their value to 0 on /myself, but I feel this will be brittle - the quality could be set to 0 without the tracker picking it up. Maybe I just need to say it's not compatible with hidden qualities?

> What I can suggest is to download data about all the qualities (name, image,
I looked at downloading this stuff from the wiki, but their icon names don't always match the game's names, so I'd need to get the names from the game itself.

> Could it be done in a way that makes specific targets optional?
I was thinking for this I'd make it so if the target is 0 it just doesn't show the '/ target' or the progress bar... or actually maybe leave the progress bar filled in, so it takes the same space and keeps things lined up.

AnteIndustrial commented:
I've changed it so qualities can have multiple targets, or zero. I'm not really sure I like it though, the edit modal is so busy now.
![image](https://github.com/user-attachments/assets/5cdcea01-fbf4-4a48-8a12-73c27a151862)
The main tracker panel looks alright though
![image](https://github.com/user-attachments/assets/1190d8f7-f395-4b50-bd63-e695f4bcc3b9)
And here's Subtle without a target
![image](https://github.com/user-attachments/assets/114215d7-e497-4643-8e42-a3b53b1bf43a)

I haven't updated the pull request because there are still some bugs and the html-adding part has ballooned, I'm trying to cut it down a bit. Let me know what you think, it's really just the edit modal I'm not happy with.
![image](https://github.com/user-attachments/assets/089fe756-7a5d-49de-ad47-10a55362f5ba)

AnteIndustrial committed:
added an 'update settings' message, so components can update a setting without needing to know the whole settings object
fixed a couple of bugs where settings and html would get out of sync
when new options are added to the dropdown, put them in order. stop adding duplicates
allow multiple, or 0, targets for any quality
